proposals/lip-0044.md

```diff
@@ -190,234 +216,320 @@ genesisDataSchema = {
+    emitEvent(
+        moduleID = MODULE_ID_VALIDATORS,
+        typeID = TYPE_ID_BLS_KEY_REGISTRATION,
+        data = {
+            "blsKey": blsKey,
+            "result": KEY_REG_RESULT_SUCCESS
+        },
+        topics=[validatorAddress]
+    )
 ```
 
 #### setValidatorGeneratorKey
 
 This function sets the generator key of a validator account.
-
-##### Parameters
-
-This function has the following input parameters in the order given below:
-
-* `validatorAddress`: A 20 bytes value identifying the validator account in the validators data substore.
-* `generatorKey`: A 32 bytes value, specifying the new generator key of the validator.
-
-##### Returns
-
-* `success`: A boolean, indicating success or failure of the function execution.
-
-##### Execution
-
-This function checks that there exists a validator account registered for the input `validatorAddress` and then updates the generator key.
+It checks that there exists a validator account registered for the input `validatorAddress` and then updates the generator key to the input `generatorKey`.
 
 ```python
-setValidatorGeneratorKey(validatorAddress, generatorKey):
+def setValidatorGeneratorKey(validatorAddress: Address, generatorKey: PublicKeyEd25519) -> None:
     if no entry in the validators data substore exist with storeKey == validatorAddress:
-        return False
+        emitPersistentEvent(
+            moduleID = MODULE_ID_VALIDATORS,
+            typeID = TYPE_ID_GENERATOR_KEY_REGISTRATION,
+            data = {
+                "generatorKey": generatorKey,
+                "result": KEY_REG_RESULT_NO_VALIDATOR
+            },
+            topics=[validatorAddress]
+        )
+        raise Exception('This address is registered as validator. Only validators can register a generator key.')
 
     validatorAccount(validatorAddress).generatorKey = generatorKey
-    return True
+
+    emitEvent(
+        moduleID = MODULE_ID_VALIDATORS,
+        typeID = TYPE_ID_GENERATOR_KEY_REGISTRATION,
+        data = {
+            "generatorKey": generatorKey,
+            "result": KEY_REG_RESULT_SUCCESS
+        },
+        topics=[validatorAddress]
+    )
 ```
 
 #### getAddressFromBLSKey
 
-This function returns the address that registered the input BLS key.
-
-##### Parameters
-
-`blsKey`: A 48 bytes value corresponding to a BLS key.
-
-##### Returns
-
-This function returns a 20 bytes address.
-
-##### Execution
-
-This function checks if an entry with store key equal to  `blsKey` exists in the registered BLS keys substore. If this is the case, it returns the stored `address`, else it returns an error.
+This function returns the address that registered the input BLS key `blsKey: BLSKey`.
+It checks if an entry with store key equal to `blsKey` exists in the registered BLS keys substore. 
+If this is the case, it returns the stored `address: Address`, else it throws an error.
 
 #### getGeneratorList
```

Comment:
Just to have an explicit input and output type, I would add some pseudocode with type-hinting here as well.